File: intelitor-v5.2/lib/cepaf/src/Cepaf.Database/HolonDatabase.fs
/// Unified Database Access for F# Holons.
///
/// WHAT: Single entry point for all F# holon database operations with direct
///       high-performance access to SQLite and DuckDB using MailboxProcessor.
///
/// WHY: SC-XHOLON-001 requires isolated database access per holon.
///      SC-XHOLON-002 mandates native high-performance libraries.
///
/// CONSTRAINTS:
///   - SC-XHOLON-001: Each holon has isolated database files
///   - SC-XHOLON-002: Direct access via Microsoft.Data.Sqlite/DuckDB.NET
///   - SC-XHOLON-010: Lock-free reads via OCC
///   - SC-XHOLON-020: SQLite read latency < 1ms
///   - SC-XHOLON-021: DuckDB query latency < 10ms
///   - SC-DBNAME-001: UHI-based database paths
///
/// ## Architecture
///
/// ```
/// HolonDatabase (MailboxProcessor)
///      │
///      ├── SQLitePool (ConnectionPool + Microsoft.Data.Sqlite)
///      │   └── state.sqlite, vectors.sqlite
///      │
///      └── DuckDBPool (ConnectionPool + DuckDB.NET)
///          └── analytics.duckdb, history.duckdb, register.duckdb
/// ```
///
/// ## Usage
///
/// ```fsharp
/// // Start database for a holon
/// let db = HolonDatabase.create "fs:l4:prj:agt:cockpit"
///
/// // Query SQLite
/// let! rows = db.Query(State, "SELECT * FROM config WHERE key = ?", ["setting"])
///
/// // Execute with version control
/// let! result = db.ExecuteCas(State, "INSERT INTO config VALUES (?, ?)", ["k"; "v"], expectedVersion)
/// ```
module Cepaf.Database.HolonDatabase

open System
open System.Collections.Generic
open System.Data
open System.IO
open Microsoft.Data.Sqlite
open Cepaf.Database.Types
open Cepaf.Database.ConnectionPool
open Cepaf.Database.HolonConcurrencyHandler
module DP = Cepaf.Holon.DatabasePath

// ============================================================================
// Types
// ============================================================================

/// Database type for holon access
type HolonDbType =
    | State     // state.sqlite (SQLite)
    | Vectors   // vectors.sqlite (SQLite)
    | Cache     // cache.sqlite (SQLite)
    | Analytics // analytics.duckdb (DuckDB)
    | History   // history.duckdb (DuckDB)
    | Register  // register.duckdb (DuckDB)

/// Query result as list of dictionaries
type QueryResult = IReadOnlyList<IReadOnlyDictionary<string, obj>>

/// Execute result
type ExecuteResult = {
    Changes: int
    LastInsertId: int64 option
    Version: VersionVector option
}

/// Database message for MailboxProcessor
type private HolonDbMessage =
    | Query of HolonDbType * string * obj list * AsyncReplyChannel<Result<QueryResult, string>>
    | Execute of HolonDbType * string * obj list * AsyncReplyChannel<Result<ExecuteResult, string>>
    | ExecuteCas of HolonDbType * string * obj list * VersionVector * AsyncReplyChannel<Result<ExecuteResult, string>>
    | Transaction of HolonDbType * (IDbConnection -> Result<obj, string>) * AsyncReplyChannel<Result<obj, string>>
    | GetVersionVector of AsyncReplyChannel<VersionVector>
    | GetStats of AsyncReplyChannel<HolonDbStats>
    | Shutdown of AsyncReplyChannel<unit>

/// Database statistics
and HolonDbStats = {
    Queries: int64
    Executes: int64
    Transactions: int64
    Conflicts: int64
    Errors: int64
    StartedAt: DateTime
    HolonId: string
}

// ============================================================================
// Configuration
// ============================================================================

/// Configuration for holon database
type HolonDbConfig = {
    HolonId: string
    PoolSize: int
    AcquireTimeout: TimeSpan
}

let defaultConfig holonId = {
    HolonId = holonId
    PoolSize = 5
    AcquireTimeout = TimeSpan.FromSeconds(5.0)
}

// ============================================================================
// SQLite Database Types
// ============================================================================

let private sqliteDbs = [State; Vectors; Cache]
let private duckdbDbs = [Analytics; History; Register]

let private dbTypeToFileType = function
    | State -> DP.DatabaseType.State
    | Vectors -> DP.DatabaseType.Vectors
    | Cache -> DP.DatabaseType.State  // Use state type for cache
    | Analytics -> DP.DatabaseType.Analytics
    | History -> DP.DatabaseType.History
    | Register -> DP.DatabaseType.Register

// ============================================================================
// Connection Helpers
// ============================================================================

/// Create SQLite connection factory
let private createSqliteConnection (path: string) () =
    Directory.CreateDirectory(Path.GetDirectoryName(path)) |> ignore
    let conn = new SqliteConnection($"Data Source={path};Mode=ReadWriteCreate")
    conn.Open()
    // Enable WAL mode for concurrent access
    use cmd = conn.CreateCommand()
    cmd.CommandText <- "PRAGMA journal_mode=WAL; PRAGMA busy_timeout=5000; PRAGMA foreign_keys=ON; PRAGMA synchronous=NORMAL"
    cmd.ExecuteNonQuery() |> ignore
    conn

/// Validate SQLite connection
let private validateSqliteConnection (conn: SqliteConnection) =
    try conn.State = ConnectionState.Open with _ -> false

/// Create DuckDB connection factory (placeholder - uses IDbConnection interface)
let private createDuckDbConnection (path: string) () =
    Directory.CreateDirectory(Path.GetDirectoryName(path)) |> ignore
    // Note: In production, use DuckDB.NET.Data.DuckDBConnection
    // For now, using placeholder that can be swapped
    let conn = new SqliteConnection($"Data Source={path}") :> IDbConnection
    conn.Open()
    conn

/// Validate DuckDB connection
let private validateDuckDbConnection (conn: IDbConnection) =
    try conn.State = ConnectionState.Open with _ -> false

// ============================================================================
// HolonDatabase Implementation
// ============================================================================

/// Holon database instance
type HolonDatabase private (config: HolonDbConfig, agent: MailboxProcessor<HolonDbMessage>) =

    /// Query the database
    member _.Query(dbType: HolonDbType, sql: string, ?parameters: obj list) : Async<Result<QueryResult, string>> =
        let params = defaultArg parameters []
        agent.PostAndAsyncReply(fun reply -> Query(dbType, sql, params, reply))

    /// Execute a write statement
    member _.Execute(dbType: HolonDbType, sql: string, ?parameters: obj list) : Async<Result<ExecuteResult, string>> =
        let params = defaultArg parameters []
        agent.PostAndAsyncReply(fun reply -> Execute(dbType, sql, params, reply))

    /// Execute with compare-and-swap (optimistic concurrency)
    member _.ExecuteCas(dbType: HolonDbType, sql: string, parameters: obj list, expectedVersion: VersionVector) : Async<Result<ExecuteResult, string>> =
        agent.PostAndAsyncReply(fun reply -> ExecuteCas(dbType, sql, parameters, expectedVersion, reply))

    /// Execute within a transaction
    member _.Transaction<'T>(dbType: HolonDbType, operation: IDbConnection -> Result<'T, string>) : Async<Result<'T, string>> = async {
        let boxedOp = fun conn ->
            match operation conn with
            | Ok result -> Ok (box result)
            | Error e -> Error e
        let! result = agent.PostAndAsyncReply(fun reply -> Transaction(dbType, boxedOp, reply))
        return
            match result with
            | Ok boxed -> Ok (unbox<'T> boxed)
            | Error e -> Error e
    }

    /// Get current version vector
    member _.GetVersionVector() : Async<VersionVector> =
        agent.PostAndAsyncReply(fun reply -> GetVersionVector reply)

    /// Get database statistics
    member _.GetStats() : Async<HolonDbStats> =
        agent.PostAndAsyncReply(fun reply -> GetStats reply)

    /// Shutdown the database
    member _.Shutdown() : Async<unit> =
        agent.PostAndAsyncReply(fun reply -> Shutdown reply)

    /// Holon ID
    member _.HolonId = config.HolonId

    /// Create a new HolonDatabase instance
    static member Create(holonId: string, ?poolSize: int, ?acquireTimeout: TimeSpan) : HolonDatabase =
        let config = {
            HolonId = holonId
            PoolSize = defaultArg poolSize 5
            AcquireTimeout = defaultArg acquireTimeout (TimeSpan.FromSeconds(5.0))
        }

        // Parse UHI
        let uhi =
            match DP.parseUHI holonId with
            | DP.Success u -> u
            | DP.Error e -> failwith $"Invalid holon ID: {e}"

        // Create pool configuration
        let poolConfig = {
            MaxConnections = config.PoolSize
            MinConnections = 1
            AcquireTimeout = config.AcquireTimeout
            IdleTimeout = TimeSpan.FromMinutes(5.0)
            ValidationQuery = Some "SELECT 1"
        }

        // Create SQLite pools
        let sqlitePools =
            sqliteDbs
            |> List.map (fun dbType ->
                let fqdn = DP.createFQDN uhi (dbTypeToFileType dbType)
                let path = DP.resolve fqdn
                let pool = new ConnectionPool<SqliteConnection>(
                    poolConfig,
                    createSqliteConnection path,
                    validateSqliteConnection
                )
                dbType, pool
            )
            |> dict

        // Create DuckDB pools (placeholder)
        let duckdbPools =
            duckdbDbs
            |> List.map (fun dbType ->
                let fqdn = DP.createFQDN uhi (dbTypeToFileType dbType)
                let path = DP.resolve fqdn
                let pool = new ConnectionPool<IDbConnection>(
                    poolConfig,
                    createDuckDbConnection path,
                    validateDuckDbConnection
                )
                dbType, pool
            )
            |> dict

        // Initialize state
        let mutable versionVectors = Map.ofList [
            "local", 0L
            holonId, 0L
        ]
        let mutable stats = {
            Queries = 0L
            Executes = 0L
            Transactions = 0L
            Conflicts = 0L
            Errors = 0L
            StartedAt = DateTime.UtcNow
            HolonId = holonId
        }

        // Helper to get pool
        let getPool dbType =
            if List.contains dbType sqliteDbs then
                match sqlitePools.TryGetValue(dbType) with
                | true, pool -> Some (Choice1Of2 pool)
                | false, _ -> None
            elif List.contains dbType duckdbDbs then
                match duckdbPools.TryGetValue(dbType) with
                | true, pool -> Some (Choice2Of2 pool)
                | false, _ -> None
            else
                None

        // Execute query helper
        let executeQuery dbType sql (parameters: obj list) = async {
            match getPool dbType with
            | None -> return Error $"Unknown database type: {dbType}"
            | Some poolChoice ->
                match poolChoice with
                | Choice1Of2 sqlitePool ->
                    let! maybeConn = sqlitePool.AcquireAsync(config.AcquireTimeout)
                    match maybeConn with
                    | None -> return Error "Failed to acquire SQLite connection"
                    | Some pooled ->
                        try
                            use cmd = pooled.Connection.CreateCommand()
                            cmd.CommandText <- sql

                            // Add parameters
                            for i, param in parameters |> List.indexed do
                                cmd.Parameters.AddWithValue($"@p{i}", param) |> ignore

                            use reader = cmd.ExecuteReader()
                            let results = List<IReadOnlyDictionary<string, obj>>()

                            while reader.Read() do
                                let row = Dictionary<string, obj>()
                                for i in 0 .. reader.FieldCount - 1 do
                                    let name = reader.GetName(i)
                                    let value = reader.GetValue(i)
                                    row.[name] <- if value :? DBNull then null else value
                                results.Add(row :> IReadOnlyDictionary<string, obj>)

                            sqlitePool.Release(pooled)
                            stats <- { stats with Queries = stats.Queries + 1L }
                            return Ok (results :> QueryResult)
                        with ex ->
                            sqlitePool.Release(pooled)
                            stats <- { stats with Errors = stats.Errors + 1L }
                            return Error $"Query failed: {ex.Message}"

                | Choice2Of2 duckdbPool ->
                    let! maybeConn = duckdbPool.AcquireAsync(config.AcquireTimeout)
                    match maybeConn with
                    | None -> return Error "Failed to acquire DuckDB connection"
                    | Some pooled ->
                        try
                            use cmd = pooled.Connection.CreateCommand()
                            cmd.CommandText <- sql

                            // Add parameters
                            for i, param in parameters |> List.indexed do
                                let p = cmd.CreateParameter()
                                p.ParameterName <- $"@p{i}"
                                p.Value <- param
                                cmd.Parameters.Add(p) |> ignore

                            use reader = cmd.ExecuteReader()
                            let results = List<IReadOnlyDictionary<string, obj>>()

                            while reader.Read() do
                                let row = Dictionary<string, obj>()
                                for i in 0 .. reader.FieldCount - 1 do
                                    let name = reader.GetName(i)
                                    let value = reader.GetValue(i)
                                    row.[name] <- if value :? DBNull then null else value
                                results.Add(row :> IReadOnlyDictionary<string, obj>)

                            duckdbPool.Release(pooled)
                            stats <- { stats with Queries = stats.Queries + 1L }
                            return Ok (results :> QueryResult)
                        with ex ->
                            duckdbPool.Release(pooled)
                            stats <- { stats with Errors = stats.Errors + 1L }
                            return Error $"Query failed: {ex.Message}"
        }

        // Execute write helper
        let executeWrite dbType sql (parameters: obj list) = async {
            match getPool dbType with
            | None -> return Error $"Unknown database type: {dbType}"
            | Some poolChoice ->
                match poolChoice with
                | Choice1Of2 sqlitePool ->
                    let! maybeConn = sqlitePool.AcquireAsync(config.AcquireTimeout)
                    match maybeConn with
                    | None -> return Error "Failed to acquire SQLite connection"
                    | Some pooled ->
                        try
                            use cmd = pooled.Connection.CreateCommand()
                            cmd.CommandText <- sql

                            // Add parameters
                            for i, param in parameters |> List.indexed do
                                cmd.Parameters.AddWithValue($"@p{i}", param) |> ignore

                            let changes = cmd.ExecuteNonQuery()

                            // Get last insert ID if applicable
                            use idCmd = pooled.Connection.CreateCommand()
                            idCmd.CommandText <- "SELECT last_insert_rowid()"
                            let lastId = idCmd.ExecuteScalar() |> Convert.ToInt64

                            // Update version
                            versionVectors <- increment versionVectors holonId

                            sqlitePool.Release(pooled)
                            stats <- { stats with Executes = stats.Executes + 1L }
                            return Ok {
                                Changes = changes
                                LastInsertId = Some lastId
                                Version = Some versionVectors
                            }
                        with ex ->
                            sqlitePool.Release(pooled)
                            stats <- { stats with Errors = stats.Errors + 1L }
                            return Error $"Execute failed: {ex.Message}"

                | Choice2Of2 duckdbPool ->
                    let! maybeConn = duckdbPool.AcquireAsync(config.AcquireTimeout)
                    match maybeConn with
                    | None -> return Error "Failed to acquire DuckDB connection"
                    | Some pooled ->
                        try
                            use cmd = pooled.Connection.CreateCommand()
                            cmd.CommandText <- sql

                            // Add parameters
                            for i, param in parameters |> List.indexed do
                                let p = cmd.CreateParameter()
                                p.ParameterName <- $"@p{i}"
                                p.Value <- param
                                cmd.Parameters.Add(p) |> ignore

                            let changes = cmd.ExecuteNonQuery()

                            // Update version
                            versionVectors <- increment versionVectors holonId

                            duckdbPool.Release(pooled)
                            stats <- { stats with Executes = stats.Executes + 1L }
                            return Ok {
                                Changes = changes
                                LastInsertId = None
                                Version = Some versionVectors
                            }
                        with ex ->
                            duckdbPool.Release(pooled)
                            stats <- { stats with Errors = stats.Errors + 1L }
                            return Error $"Execute failed: {ex.Message}"
        }

        // Create the MailboxProcessor agent
        let agent = MailboxProcessor.Start(fun inbox ->
            let rec loop () = async {
                let! msg = inbox.Receive()

                match msg with
                | Query (dbType, sql, parameters, reply) ->
                    let! result = executeQuery dbType sql parameters
                    reply.Reply(result)
                    return! loop()

                | Execute (dbType, sql, parameters, reply) ->
                    let! result = executeWrite dbType sql parameters
                    reply.Reply(result)
                    return! loop()

                | ExecuteCas (dbType, sql, parameters, expectedVersion, reply) ->
                    // Check version before executing
                    if versionGte versionVectors expectedVersion then
                        let! result = executeWrite dbType sql parameters
                        reply.Reply(result)
                    else
                        stats <- { stats with Conflicts = stats.Conflicts + 1L }
                        reply.Reply(Error "conflict")
                    return! loop()

                | Transaction (dbType, operation, reply) ->
                    match getPool dbType with
                    | None ->
                        reply.Reply(Error $"Unknown database type: {dbType}")
                        return! loop()
                    | Some poolChoice ->
                        match poolChoice with
                        | Choice1Of2 sqlitePool ->
                            let! maybeConn = sqlitePool.AcquireAsync(config.AcquireTimeout)
                            match maybeConn with
                            | None ->
                                reply.Reply(Error "Failed to acquire connection")
                                return! loop()
                            | Some pooled ->
                                try
                                    try
                                        use txn = pooled.Connection.BeginTransaction()
                                        let conn = pooled.Connection :> IDbConnection
                                        match operation conn with
                                        | Ok result ->
                                            txn.Commit()
                                            versionVectors <- increment versionVectors holonId
                                            stats <- { stats with Transactions = stats.Transactions + 1L }
                                            reply.Reply(Ok result)
                                        | Error e ->
                                            txn.Rollback()
                                            stats <- { stats with Errors = stats.Errors + 1L }
                                            reply.Reply(Error e)
                                    with ex ->
                                        stats <- { stats with Errors = stats.Errors + 1L }
                                        reply.Reply(Error $"Transaction failed: {ex.Message}")
                                finally
                                    sqlitePool.Release(pooled)
                                return! loop()

                        | Choice2Of2 duckdbPool ->
                            let! maybeConn = duckdbPool.AcquireAsync(config.AcquireTimeout)
                            match maybeConn with
                            | None ->
                                reply.Reply(Error "Failed to acquire connection")
                                return! loop()
                            | Some pooled ->
                                try
                                    try
                                        // DuckDB has limited transaction support
                                        let conn = pooled.Connection
                                        match operation conn with
                                        | Ok result ->
                                            versionVectors <- increment versionVectors holonId
                                            stats <- { stats with Transactions = stats.Transactions + 1L }
                                            reply.Reply(Ok result)
                                        | Error e ->
                                            stats <- { stats with Errors = stats.Errors + 1L }
                                            reply.Reply(Error e)
                                    with ex ->
                                        stats <- { stats with Errors = stats.Errors + 1L }
                                        reply.Reply(Error $"Transaction failed: {ex.Message}")
                                finally
                                    duckdbPool.Release(pooled)
                                return! loop()

                | GetVersionVector reply ->
                    reply.Reply(versionVectors)
                    return! loop()

                | GetStats reply ->
                    reply.Reply(stats)
                    return! loop()

                | Shutdown reply ->
                    // Dispose all pools
                    for kvp in sqlitePools do
                        (kvp.Value :> IDisposable).Dispose()
                    for kvp in duckdbPools do
                        (kvp.Value :> IDisposable).Dispose()
                    reply.Reply(())
                    // Don't loop, terminate
            }

            loop()
        )

        printfn "[HolonDatabase] Initialized database for holon: %s" holonId
        HolonDatabase(config, agent)

// ============================================================================
// Convenience Functions
// ============================================================================

/// Create a new holon database with default settings
let create holonId = HolonDatabase.Create(holonId)

/// Create a new holon database with custom settings
let createWithConfig holonId poolSize acquireTimeout =
    HolonDatabase.Create(holonId, poolSize, acquireTimeout)

// ============================================================================
// Telemetry
// ============================================================================

/// Emit telemetry event (placeholder for Zenoh integration)
let private emitTelemetry (operation: string) (holonId: string) (dbType: HolonDbType) (durationUs: int64) (success: bool) =
    printfn "[HolonDatabase] %s %s %A %dμs %s"
        operation
        holonId
        dbType
        durationUs
        (if success then "OK" else "ERROR")

File: sub-projects/intelitor-v5.2/lib/indrajaal/holon/database/holon_database.ex
defmodule Indrajaal.Holon.Database.HolonDatabase do
  @moduledoc """
  Unified Database Access for Elixir Holons.

  WHAT: Single entry point for all holon database operations with direct
        high-performance access to SQLite and DuckDB.

  WHY: SC-XHOLON-001 requires isolated database access per holon.
       SC-XHOLON-002 mandates native high-performance libraries.

  CONSTRAINTS:
    - SC-XHOLON-001: Each holon has isolated database files
    - SC-XHOLON-002: Direct access via Exqlite/Duckdbex
    - SC-XHOLON-010: Lock-free reads via OCC
    - SC-XHOLON-020: SQLite read latency < 1ms
    - SC-XHOLON-021: DuckDB query latency < 10ms
    - SC-DBNAME-001: UHI-based database paths

  ## Architecture

  ```
  HolonDatabase (GenServer)
       │
       ├── SQLitePool (Poolboy + Exqlite)
       │   └── state.sqlite, vectors.sqlite, cache.sqlite
       │
       └── DuckDBPool (Poolboy + Duckdbex)
           └── analytics.duckdb, history.duckdb, register.duckdb
  ```

  ## Usage

  ```elixir
  # Start database for a holon
  {:ok, pid} = HolonDatabase.start_link(holon_id: "ex:l3:kms:srv:main")

  # Read from SQLite
  {:ok, rows} = HolonDatabase.query(:state, "SELECT * FROM config WHERE key = ?", ["setting"])

  # Write to SQLite with transaction
  {:ok, _} = HolonDatabase.transaction(:state, fn conn ->
    HolonDatabase.execute(conn, "INSERT INTO config (key, value) VALUES (?, ?)", ["k", "v"])
  end)

  # Query DuckDB for analytics
  {:ok, rows} = HolonDatabase.query(:analytics, "SELECT * FROM metrics WHERE ts > ?", [timestamp])
  ```
  """

  use GenServer
  require Logger

  alias Indrajaal.Holon.DatabasePath
  alias Indrajaal.Holon.Database.{SQLitePool, DuckDBPool, ConcurrencyHandler}

  @type holon_id :: String.t()
  @type db_type :: :state | :analytics | :history | :vectors | :register | :cache
  @type query_result :: {:ok, [map()]} | {:error, String.t()}
  @type version_vector :: %{String.t() => non_neg_integer()}

  @sqlite_dbs [:state, :vectors, :cache]
  @duckdb_dbs [:analytics, :history, :register]

  defstruct [
    :holon_id,
    :sqlite_pools,
    :duckdb_pools,
    :version_vectors,
    :stats
  ]

  # ============================================================================
  # Client API
  # ============================================================================

  @doc """
  Start a HolonDatabase for the given holon.

  ## Options
    - `:holon_id` - Required. The UHI of the holon (e.g., "ex:l3:kms:srv:main")
    - `:pool_size` - Optional. Connection pool size (default: 5)
    - `:acquire_timeout` - Optional. Pool acquire timeout in ms (default: 5000)
  """
  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts) do
    holon_id = Keyword.fetch!(opts, :holon_id)
    name = via_tuple(holon_id)
    GenServer.start_link(__MODULE__, opts, name: name)
  end

  @doc """
  Execute a read query on the specified database type.

  ## Parameters
    - `holon_id` - The UHI of the holon
    - `db_type` - One of :state, :vectors, :cache (SQLite) or :analytics, :history, :register (DuckDB)
    - `sql` - The SQL query string
    - `params` - Query parameters (default: [])

  ## Returns
    - `{:ok, rows}` on success
    - `{:error, reason}` on failure

  ## Example
      {:ok, rows} = HolonDatabase.query("ex:l3:kms:srv:main", :state, "SELECT * FROM config")
  """
  @spec query(holon_id(), db_type(), String.t(), list()) :: query_result()
  def query(holon_id, db_type, sql, params \\ []) do
    GenServer.call(via_tuple(holon_id), {:query, db_type, sql, params})
  end

  @doc """
  Execute a write statement on the specified database type.

  ## Parameters
    - `holon_id` - The UHI of the holon
    - `db_type` - Database type
    - `sql` - The SQL statement
    - `params` - Statement parameters

  ## Returns
    - `{:ok, %{changes: n}}` on success
    - `{:error, reason}` on failure
  """
  @spec execute(holon_id(), db_type(), String.t(), list()) :: {:ok, map()} | {:error, String.t()}
  def execute(holon_id, db_type, sql, params \\ []) do
    GenServer.call(via_tuple(holon_id), {:execute, db_type, sql, params})
  end

  @doc """
  Execute multiple statements within a transaction.

  ## Parameters
    - `holon_id` - The UHI of the holon
    - `db_type` - Database type (SQLite only for full ACID)
    - `fun` - Function receiving connection, should return {:ok, result} or {:error, reason}
    - `opts` - Transaction options
      - `:isolation` - :serializable, :repeatable_read, :read_committed (default: :serializable)
      - `:timeout` - Transaction timeout in ms (default: 5000)

  ## Example
      {:ok, result} = HolonDatabase.transaction("ex:l3:kms:srv:main", :state, fn conn ->
        HolonDatabase.conn_execute(conn, "INSERT INTO t VALUES (?)", [1])
        HolonDatabase.conn_execute(conn, "INSERT INTO t VALUES (?)", [2])
        {:ok, :inserted}
      end)
  """
  @spec transaction(
          holon_id(),
          db_type(),
          (term() -> {:ok, term()} | {:error, term()}),
          keyword()
        ) ::
          {:ok, term()} | {:error, term()}
  def transaction(holon_id, db_type, fun, opts \\ []) do
    GenServer.call(via_tuple(holon_id), {:transaction, db_type, fun, opts})
  end

  @doc """
  Get current version vector for optimistic concurrency control.
  """
  @spec get_version_vector(holon_id()) :: {:ok, version_vector()}
  def get_version_vector(holon_id) do
    GenServer.call(via_tuple(holon_id), :get_version_vector)
  end

  @doc """
  Execute with optimistic concurrency control (compare-and-swap).

  ## Parameters
    - `holon_id` - The UHI of the holon
    - `db_type` - Database type
    - `sql` - The SQL statement
    - `params` - Statement parameters
    - `expected_version` - Expected version vector

  ## Returns
    - `{:ok, %{changes: n, new_version: vv}}` on success
    - `{:conflict, current_version}` on version mismatch
    - `{:error, reason}` on failure
  """
  @spec execute_cas(holon_id(), db_type(), String.t(), list(), version_vector()) ::
          {:ok, map()} | {:conflict, version_vector()} | {:error, String.t()}
  def execute_cas(holon_id, db_type, sql, params, expected_version) do
    GenServer.call(via_tuple(holon_id), {:execute_cas, db_type, sql, params, expected_version})
  end

  @doc """
  Get database statistics.
  """
  @spec stats(holon_id()) :: {:ok, map()}
  def stats(holon_id) do
    GenServer.call(via_tuple(holon_id), :stats)
  end

  # ============================================================================
  # GenServer Callbacks
  # ============================================================================

  @impl true
  def init(opts) do
    holon_id = Keyword.fetch!(opts, :holon_id)
    pool_size = Keyword.get(opts, :pool_size, 5)
    acquire_timeout = Keyword.get(opts, :acquire_timeout, 5_000)

    Logger.info("[HolonDatabase] Initializing database for holon: #{holon_id}")

    # Initialize SQLite pools
    sqlite_pools =
      @sqlite_dbs
      |> Enum.map(fn db_type ->
        fqdn = "#{holon_id}:#{db_type}"
        {:ok, path} = DatabasePath.resolve(fqdn)
        ensure_directory(path)

        pool_name = pool_name(holon_id, db_type)
        {:ok, _} = SQLitePool.start_pool(pool_name, path, pool_size, acquire_timeout)
        {db_type, pool_name}
      end)
      |> Map.new()

    # Initialize DuckDB pools
    duckdb_pools =
      @duckdb_dbs
      |> Enum.map(fn db_type ->
        fqdn = "#{holon_id}:#{db_type}"
        {:ok, path} = DatabasePath.resolve(fqdn)
        ensure_directory(path)

        pool_name = pool_name(holon_id, db_type)
        {:ok, _} = DuckDBPool.start_pool(pool_name, path, pool_size, acquire_timeout)
        {db_type, pool_name}
      end)
      |> Map.new()

    # Initialize version vectors
    version_vectors = %{
      "local" => 0,
      holon_id => 0
    }

    # Initialize statistics
    stats = %{
      queries: 0,
      executes: 0,
      transactions: 0,
      conflicts: 0,
      errors: 0,
      started_at: DateTime.utc_now()
    }

    state = %__MODULE__{
      holon_id: holon_id,
      sqlite_pools: sqlite_pools,
      duckdb_pools: duckdb_pools,
      version_vectors: version_vectors,
      stats: stats
    }

    {:ok, state}
  end

  @impl true
  def handle_call({:query, db_type, sql, params}, _from, state) do
    start_time = System.monotonic_time(:microsecond)

    result =
      case get_pool(state, db_type) do
        {:sqlite, pool_name} ->
          SQLitePool.query(pool_name, sql, params)

        {:duckdb, pool_name} ->
          DuckDBPool.query(pool_name, sql, params)

        {:error, reason} ->
          {:error, reason}
      end

    duration = System.monotonic_time(:microsecond) - start_time
    emit_telemetry(:query, state.holon_id, db_type, duration, result)

    new_stats = update_stats(state.stats, :queries, result)
    {:reply, result, %{state | stats: new_stats}}
  end

  @impl true
  def handle_call({:execute, db_type, sql, params}, _from, state) do
    start_time = System.monotonic_time(:microsecond)

    result =
      case get_pool(state, db_type) do
        {:sqlite, pool_name} ->
          SQLitePool.execute(pool_name, sql, params)

        {:duckdb, pool_name} ->
          DuckDBPool.execute(pool_name, sql, params)

        {:error, reason} ->
          {:error, reason}
      end

    # Update version vector on successful write
    {result, new_vv} =
      case result do
        {:ok, changes} ->
          new_vv = increment_version(state.version_vectors, state.holon_id)
          {{:ok, Map.put(changes, :version, new_vv)}, new_vv}

        error ->
          {error, state.version_vectors}
      end

    duration = System.monotonic_time(:microsecond) - start_time
    emit_telemetry(:execute, state.holon_id, db_type, duration, result)

    new_stats = update_stats(state.stats, :executes, result)
    {:reply, result, %{state | stats: new_stats, version_vectors: new_vv}}
  end

  @impl true
  def handle_call({:transaction, db_type, fun, opts}, _from, state) do
    start_time = System.monotonic_time(:microsecond)

    result =
      case get_pool(state, db_type) do
        {:sqlite, pool_name} ->
          SQLitePool.transaction(pool_name, fun, opts)

        {:duckdb, _pool_name} ->
          # DuckDB has limited transaction support
          {:error, "DuckDB transactions not fully supported, use SQLite for ACID"}

        {:error, reason} ->
          {:error, reason}
      end

    # Update version vector on successful transaction
    {result, new_vv} =
      case result do
        {:ok, _} = ok ->
          new_vv = increment_version(state.version_vectors, state.holon_id)
          {ok, new_vv}

        error ->
          {error, state.version_vectors}
      end

    duration = System.monotonic_time(:microsecond) - start_time
    emit_telemetry(:transaction, state.holon_id, db_type, duration, result)

    new_stats = update_stats(state.stats, :transactions, result)
    {:reply, result, %{state | stats: new_stats, version_vectors: new_vv}}
  end

  @impl true
  def handle_call({:execute_cas, db_type, sql, params, expected_version}, _from, state) do
    # Compare version vectors
    if ConcurrencyHandler.version_gte?(state.version_vectors, expected_version) do
      # Version matches, proceed with execute
      result =
        case get_pool(state, db_type) do
          {:sqlite, pool_name} ->
            SQLitePool.execute(pool_name, sql, params)

          {:duckdb, pool_name} ->
            DuckDBPool.execute(pool_name, sql, params)

          {:error, reason} ->
            {:error, reason}
        end

      case result do
        {:ok, changes} ->
          new_vv = increment_version(state.version_vectors, state.holon_id)
          new_stats = update_stats(state.stats, :executes, result)

          {:reply, {:ok, Map.put(changes, :new_version, new_vv)},
           %{state | stats: new_stats, version_vectors: new_vv}}

        error ->
          new_stats = update_stats(state.stats, :errors, error)
          {:reply, error, %{state | stats: new_stats}}
      end
    else
      # Version conflict
      new_stats = Map.update!(state.stats, :conflicts, &(&1 + 1))
      {:reply, {:conflict, state.version_vectors}, %{state | stats: new_stats}}
    end
  end

  @impl true
  def handle_call(:get_version_vector, _from, state) do
    {:reply, {:ok, state.version_vectors}, state}
  end

  @impl true
  def handle_call(:stats, _from, state) do
    {:reply, {:ok, state.stats}, state}
  end

  @impl true
  def terminate(reason, state) do
    Logger.info(
      "[HolonDatabase] Shutting down database for holon: #{state.holon_id}, reason: #{inspect(reason)}"
    )

    # Stop all pools
    Enum.each(state.sqlite_pools, fn {_type, pool_name} ->
      SQLitePool.stop_pool(pool_name)
    end)

    Enum.each(state.duckdb_pools, fn {_type, pool_name} ->
      DuckDBPool.stop_pool(pool_name)
    end)

    :ok
  end

  # ============================================================================
  # Private Functions
  # ============================================================================

  defp via_tuple(holon_id) do
    {:via, Registry, {Indrajaal.Holon.Database.Registry, holon_id}}
  end

  defp pool_name(holon_id, db_type) do
    String.to_atom("holon_db_#{String.replace(holon_id, ":", "_")}_#{db_type}")
  end

  defp get_pool(state, db_type) when db_type in @sqlite_dbs do
    case Map.fetch(state.sqlite_pools, db_type) do
      {:ok, pool_name} -> {:sqlite, pool_name}
      :error -> {:error, "Unknown SQLite database type: #{db_type}"}
    end
  end

  defp get_pool(state, db_type) when db_type in @duckdb_dbs do
    case Map.fetch(state.duckdb_pools, db_type) do
      {:ok, pool_name} -> {:duckdb, pool_name}
      :error -> {:error, "Unknown DuckDB database type: #{db_type}"}
    end
  end

  defp get_pool(_state, db_type) do
    {:error, "Unknown database type: #{db_type}"}
  end

  defp ensure_directory(path) do
    dir = Path.dirname(path)
    File.mkdir_p!(dir)
  end

  defp increment_version(version_vectors, holon_id) do
    Map.update(version_vectors, holon_id, 1, &(&1 + 1))
  end

  defp update_stats(stats, key, result) do
    stats
    |> Map.update!(key, &(&1 + 1))
    |> maybe_increment_errors(result)
  end

  defp maybe_increment_errors(stats, {:error, _}), do: Map.update!(stats, :errors, &(&1 + 1))
  defp maybe_increment_errors(stats, _), do: stats

  defp emit_telemetry(operation, holon_id, db_type, duration_us, result) do
    status = if match?({:ok, _}, result), do: :ok, else: :error

    :telemetry.execute(
      [:indrajaal, :holon, :database, operation],
      %{duration_us: duration_us},
      %{holon_id: holon_id, db_type: db_type, status: status}
    )
  end
end
